src/bincode/optional.zig

```diff
@@ -0,0 +1,48 @@
+const std = @import("std");
+const sig = @import("../lib.zig");
+const bincode = sig.bincode;
+
+pub fn defaultToNullOnEof(comptime T: type, fields: struct {
+    free: ?fn (allocator: std.mem.Allocator, data: anytype) void = null,
+    hashmap: bincode.MaybeHashMapConfig(sig.utils.types.hashMapInfo(T)) = .{},
+}) bincode.FieldConfig(?T) {
+    const gen = struct {
+        fn deserializer(
+            allocator: std.mem.Allocator,
+            reader: anytype,
+            params: bincode.Params,
+        ) anyerror!?T {
+            return bincode.read(allocator, T, reader, params) catch |err| switch (err) {
+                error.EndOfStream => null,
+                else => |e| e,
+            };
+        }
+
+        fn serializer(
+            writer: anytype,
+            maybe_data: anytype,
+            params: bincode.Params,
+        ) anyerror!void {
+            const data = maybe_data orelse return;
+            try bincode.write(writer, data, params);
+        }
+
+        fn default(_: std.mem.Allocator) ?T {
+            return null;
+        }
+
+        fn skipWrite(maybe_value: anytype) bool {
+            return maybe_value == null;
+        }
```

Comment:
do we need skipWrite if the serializer returns early in the null case?

Reply:
That's a fair question. Probably not.
I wrote this after I came up with the skip predicate - maybe it makes the predicate unnecessary, let play around with this.

Reply:
4ca848b798c075df8cde9c567f2cec9e7a1732b4

I dove in and pulled out a bunch of redundancies, this predicate included.